zebrad/src/components/mempool.rs

```diff
@@ -100,6 +146,38 @@ impl Service<Request> for Mempool {
                     .map(Response::RejectedTransactionIds);
                 async move { rsp }.boxed()
             }
+            Request::DownloadAndVerify(gossiped_txs) => {
+                let rsp = gossiped_txs
+                    .into_iter()
+                    .map(|gossiped_tx| {
+                        let r = self.should_download_or_verify(gossiped_tx.id());
+                        if let Err(action) = r {
+                            return action;
+                        }
+                        DownloadAction::DownloadAction(
+                            self.tx_downloads.download_if_needed_and_verify(gossiped_tx),
+                        )
+                    })
+                    .collect();
+                async move { Ok(Response::DownloadActions(rsp)) }.boxed()
+            }
+        }
+    }
+}
+
+impl Mempool {
+    /// Check if transaction should be downloaded and/or verified.
+    ///
+    /// If it is already in the mempool (or in its rejected list)
+    /// then it shouldn't be downloaded/verified.
+    fn should_download_or_verify(&mut self, txid: UnminedTxId) -> Result<(), DownloadAction> {
+        // Check if the transaction is already in the mempool.
+        if self.storage.clone().contains(&txid) {
+            return Err(DownloadAction::InMempool);
+        }
+        if self.storage.clone().contains_rejected(&txid) {
+            return Err(DownloadAction::Rejected);
         }
+        Ok(())
     }
 }
```

Comment:
Technically this is fine, we could consolidate into the one `impl Mempool` above

Reply:
Oops, moved in 44e7158